### ISSUE TITLE
VisionTransformer模型跟进

### ISSUE THREAD

**Issue by rentainhe** [Oneflow-Inc/libai#70]
## Vision Transformer
将在这个issue下跟进Libai VisionTransformer的各个工作进度
- 目前已经完成的工作
  - [x] Eager ViT的适配与Loss对齐，相关PR https://github.com/Oneflow-Inc/libai/pull/74
  - [x] Loss对齐的细节整理 https://github.com/Oneflow-Inc/libai/issues/63
- 后续需要继续进行的工作
  - [x] 是否需要使用最新flowvision下对齐timm的Vision Transformer模型进行Loss对齐操作 (optional): [__相关实验__](https://github.com/Oneflow-Inc/libai/issues/70#issuecomment-1019765873)
  - [x] 测试Evaluator是否兼容: 兼容Evaluator，但是Evaluator里的一个info有点问题，CP会帮忙修复～
  - [x] 使用修改后的Eager ViT，完整跑通CIFAR100
  - [x] 使用对齐的Eager ViT的模型先在Libai下进行ImageNet的完整训练，看看最终结果是否会有问题
  - [x] 使用Libai内置的Layers实现Vision Transformer后进行Loss对齐实验
  - [x] 使用Libai实现的Vision Transfor进行完整训练

## Experiments
- 简单进行了在CIFAR100上的训练: [log.txt](https://github.com/Oneflow-Inc/libai/files/7948479/log.txt)

### ImageNet
- 未完全对齐超参

| Model | Acc@1 (libai) | Acc@1 (original) |
|:-------:|:---------------:|:------------------:|
| ViT-Tiny | 71.6 | 72.2 |  

### CIFAR100
- 在这个修复的PR https://github.com/Oneflow-Inc/libai/pull/226 settings下的执行结果

| Model | Acc@1 (libai) | Acc@1 (original) |
|:-------:|:---------------:|:------------------:|
| ViT-Tiny | **69.1** | 65.5 |  

### ImageNet
- 在这个修复的PR https://github.com/Oneflow-Inc/libai/pull/226 settings下的执行结果

| Model | Acc@1 (libai) | Acc@1 (original) |
|:-------:|:---------------:|:------------------:|
| ViT-Tiny | **72.7** | 72.3 |

**Comment by rentainhe** [comment]
## 最新对齐的Vision Transformer模型的Loss对齐
- 实验超参设置 https://github.com/Oneflow-Inc/libai/issues/63

相关Loss对比图
![loss_compare](https://user-images.githubusercontent.com/48727989/150733628-660ee550-23a4-46a7-a6ce-685665e976ce.png)

- 结论：相同配置下Loss可以对齐

**Comment by rentainhe** [comment]
## DeiT实验配置
Swin-T和DeiT用的超参配置一样，这边一起记录一下，方便后续完整实验检查
- `batch size=1024`, 可以替换为8 * 128
- `lr = 5e-4`，需要根据batch size进行linear scale操作
- `optim=AdamW, Betas=(0.9, 0.999)`
- `weight decay=0.05`
- `warmup epochs=5`, `warmup epochs=20`应该也是OK的
- `label-smooth=0.1`
- `Rand Augment=9/0.5`
- `Mixup Prob=0.8`
- `Cutmix Prob=1.0`
- `Erasing Prob=0.25`

**Comment by rentainhe** [comment]
## VisionTransformer速度测试

| Model | Codebase | Framework | Batchsize | Throughput averaged over 30 times |
|:---|:---|:---|:---|---|
| vit_tiny_patch16_224 | [Swin-Transformer](https://github.com/microsoft/Swin-Transformer) | Pytorch | 64 | 2369.2 |
| vit_tiny_patch16_224 | [libai](https://github.com/Oneflow-Inc/libai) | OneFlow | 64 | - |
